apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4SamhsaMatcher.java

```diff
@@ -522,43 +523,93 @@ private static String normalizeIcd10Code(String icd10DiagnosisCode) {
-   *     {@link Coding}s that match any of the {@link #cptCodes}, <code>false</code> if they all do
-   *     not
+   *     {@link Coding}s that match any of the {@link #cptCodes} and has no unknown {@link System}s,
+   *     <code>false</code> if they all do not
+   */
+  private boolean containsHcpcsCoding(CodeableConcept procedureConcept) {
+    // Does the CodeableConcept have a legit HCPCS Coding?
+    boolean hasHcpcsCoding = findHcpcsCoding(procedureConcept);
+
+    // Check that Coding to see if it's blacklisted.
+    if (hasHcpcsCoding && isSamhsaCptCode(procedureConcept)) {
+      return true;
+    } else if (hasHcpcsCoding && !containsOnlyKnownSystems(procedureConcept)) {
+      /*
+       * Fail safe: if we don't know the procedure Coding system, assume the code is
+       * SAMHSA.
+       */
+      return true;
+    } else {
+      // Otherwise, this only has known & non-SAMHSA-blacklisted procedure codes.
+      return false;
+    }
+  }
+
+  /**
+   * @param procedureConcept the procedure {@link CodeableConcept} to check
+   * @return <code>true</code> if the specified procedure {@link CodeableConcept} contains the
+   *     {@link Coding} with the HCPCS {@link System}, <code>false</code> if it does not
    */
-  private boolean containsSamhsaProcedureCode(CodeableConcept procedureConcept) {
+  private boolean findHcpcsCoding(CodeableConcept procedureConcept) {
     for (Coding procedureCoding : procedureConcept.getCoding()) {
       if (TransformerConstants.CODING_SYSTEM_HCPCS.equals(procedureCoding.getSystem())) {
-        if (isSamhsaCptCode(procedureCoding)) return true;
-      } else {
-        /*
-         * Fail safe: if we don't know the procedure Coding system, assume the code is
-         * SAMHSA.
-         */
         return true;
       }
     }
-
-    // No blacklisted procedure Codings found: this procedure isn't SAMHSA-related.
     return false;
   }
 
   /**
-   * @param procedureCoding the procedure {@link Coding} to check
-   * @return <code>true</code> if the specified procedure {@link Coding} matches one of the {@link
-   *     #cptCodes} entries, <code>false</code> if it does not
+   * @param procedureConcept the procedure {@link CodeableConcept} to check
+   * @return <code>false</code> if the specified procedure {@link CodeableConcept} contains any
+   *     {@link Coding}s that do not contain any of the {@link System}s <code>true</code> if they do
    */
-  private boolean isSamhsaCptCode(Coding procedureCoding) {
+  private boolean containsOnlyKnownSystems(CodeableConcept procedureConcept) {
+    Set<String> codingSystems =
+        procedureConcept.getCoding().stream().map(Coding::getSystem).collect(Collectors.toSet());
+
+    String hcpcsCdSystem =
+        String.format(
+            "%s/%s",
+            TransformerConstants.BASE_URL_CCW_VARIABLES,
+            CcwCodebookVariable.HCPCS_CD.getVariable().getId().toLowerCase());
```

Comment:
makes more sense to me to use the method like you were doing before (the one that gets called from the transformer) but just move that method into a class that is not v2 specific.